File: 125-valid-palindrome/valid-palindrome.cpp
#include <cctype>

class Solution {
public:
    bool isPalindrome(string s) {
        int leftPtr = 0;
        int rightPtr = s.size();
        while (leftPtr < rightPtr) {
            char leftChar = s[leftPtr];
            char rightChar = s[rightPtr];
            if (isalnum(leftChar) && isalnum(rightChar)) {
                if (tolower(leftChar) == tolower(rightChar)) {
                    leftPtr++;
                    rightPtr--;
                    continue;
                } else {
                    return false;
                }
            }
            if (!isalnum(leftChar)) {
                leftPtr++;
            }
            if (!isalnum(rightChar)) {
                rightPtr--;
            }
        }
        return true;
    }
};

File: 125-valid-palindrome/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # preprocess first
        str_list = []
        for i in range(len(s)):
            if s[i].isalnum():
                str_list.append(str(s[i]).lower())
        s = ''.join(str_list)
        r = len(s) - 1
        for i in range(len(s) // 2):
            if s[i] != s[r]:
                return False
            r -= 1
        return True
            
            
        